src/gp_creds.c

```diff
@@ -667,6 +667,44 @@ uint32_t gp_add_krb5_creds(uint32_t *min,
             if (ret_maj) {
                 goto done;
             }
+
+            if (req_name != GSS_C_NO_NAME) {
+                int equal = 0;
+
+                ret_maj = gss_inquire_cred(&ret_min, impersonator_cred,
+                                           &target_name, NULL, NULL, NULL);
+                if (ret_maj) {
+                    goto done;
+                }
+
+                ret_maj = gss_compare_name(&ret_min, target_name,
+                                           req_name, &equal);
+                if (ret_maj) {
+                    goto done;
+                }
+
+                /* if impersonator credential retrieval yielded the requested
+                 * client name, we do not need to impersonate. Also, with MIT
+                 * krb5 an attempt to impersonate oneself gives an error "KDC
+                 * has no support for padata type" */
+                if (equal) {
+                     ret_maj = gss_acquire_cred_from(&ret_min, req_name, GSS_C_INDEFINITE,
+                                                     &desired_mechs, cred_usage,
+                                                     &cred_store, &user_cred,
+                                                     actual_mechs, NULL);
+                     if (ret_maj == GSS_S_COMPLETE) {
+                         *output_cred_handle = user_cred;
+                         user_cred = GSS_C_NO_CREDENTIAL;
+                         goto done;
+                     }
+
+                     /* fall on through, if failed */
+                 }
+
+                 gss_release_name(&discard, &target_name);
```

Comment:
this still leaks if gss_compare_name fails (unlikely but...).
I think you should simply define a new variable, near target_name, initialized to GSS_C_NO_NAME and then just release it in the done section like very other variable.
Call it compare_name please.

Reply:
All done. Thanks!